UNITTESTS/drivers/PwmOut/test_pwmout.cpp

```diff
@@ -0,0 +1,202 @@
+    }
+
+    static void release_instance()
+    {
+        --ref_counter;
+
+        if ((instance != nullptr) && (ref_counter == 0)) {
+            delete instance;
+            instance = nullptr;
+        }
+    }
+
+    MockMbedPowerMgmt(MockMbedPowerMgmt const&) = delete;
+    void operator=(MockMbedPowerMgmt const&) = delete;
+private:
+    MockMbedPowerMgmt() {}
+
+    static MockMbedPowerMgmt *instance;
+    static int ref_counter;
+};
+MockMbedPowerMgmt *MockMbedPowerMgmt::instance = nullptr;
+int MockMbedPowerMgmt::ref_counter = 0;
+
+
+void sleep_manager_unlock_deep_sleep(void)
+{
+    // Get the singleton if a test case has created one.
+    // We do not want to increment the call count unless it's done by a test
+    // case.
+    MockMbedPowerMgmt *mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(false);
+    if (mbed_mgmt_mock != nullptr) {
+        // Call only if the singleton was created by a test case.
+        mbed_mgmt_mock->sleep_manager_unlock_deep_sleep_internal();
+    }
+}
+
+
+void sleep_manager_lock_deep_sleep(void)
+{
+    // Get the singleton if a test case has created one.
+    // We do not want to increment the call count unless it's done by a test
+    // case.
+    MockMbedPowerMgmt *mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(false);
+    if (mbed_mgmt_mock != nullptr) {
+        // Call only if the singleton  was created by a test case.
+        mbed_mgmt_mock->sleep_manager_lock_deep_sleep_internal();
+    }
+}
+
+
+class TestPwmOut : public testing::Test {
+protected:
+    PwmOut *pwm_obj;
+    MockMbedPowerMgmt *mbed_mgmt_mock;
+    
+    void SetUp()
+    {
+        pwm_obj = new PwmOut(PTC1);
+        // Create a mock object singleton after the PwmOut object
+        // instantiation so the sleep_manager_lock_deep_sleep call by the
+        // constructor doesnot increment the call count.
+        // Now it is ok because a test case is about to start.
+        mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(true);
+    }
+
+    void TearDown()
+    {
+        // Destroy the mock object singleton before the PwmOut destruction
+        // because it will increment the sleep_manager_unlock_deep_sleep call
+        // count.
+        MockMbedPowerMgmt::release_instance();
+        delete pwm_obj;
+    }
+};
+// *INDENT-ON*
+
+/** Test if `suspend` unlocks deepsleep
+
+    Given an initialised Pmw with a deep sleep lock
+    When the instance is suspended
+    Then the deep sleep lock is released once
+ */
+TEST_F(TestPwmOut, test_suspend)
+{
+    using ::testing::Mock;
+
+    EXPECT_CALL(
+        *mbed_mgmt_mock, sleep_manager_unlock_deep_sleep_internal()
+    ).Times(1);
+    pwm_obj->suspend();
+    pwm_obj->suspend();
+
+    // Force gMock to verify a mock object singleton before it is destructed
+    // by the teardown
+    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
+    // There will not be a leak since the destructor releases the singleton
+    Mock::AllowLeak(mbed_mgmt_mock);
```

Comment:
This method was specifically implemented for cases where the testing framework cannot  detect that a mocked object has been safely destroyed. I destroy the singleton instance in the teardown so it is fine.

The leak error which is suppressed by `AllowLeak` occurs because I added `VerifyAndClearExpectations` so the expectation assertion does not include the call in the `PwmOut` destructor found in the teardown.

Reply:
I was confused that the comment says "There will not be a leak" and the code  "Mock::AllowLeak".
It seemed that one of them was wrong.

Perhaps you could slighlty rework  "Suppress the memory leak error. The mock singleton will be released and deleted in the TearDown call."

Reply:
[Done](https://github.com/ARMmbed/mbed-os/compare/8bcddc487357de00471d299ea00842fa12506638..6352a011849a79d8ad6d55b91d7c5d9a2e8ae1d1).